metapool/metapool.py

```diff
@@ -1458,9 +1460,419 @@ def add_syndna(plate_df, syndna_pool_number=None, syndna_concentration=None,
+                                    'RackID',
+                                    'number_of_cells']],
+            on=['LocationRow', 'RackID'],
+            how='left')
+        katharoseq_merged.rename(columns={'RackID': 'Kathseq_RackID'},
+                                 inplace=True)
+
+        # Concatenate controls into a "Controls" table and add a
+        # column named "control_sample"
+        controls = pd.concat([blanks, katharoseq_merged])
+        controls = controls.drop(['LocationCell', 'LocationColumn',
+                                  'LocationRow'], axis=1)
+
+        # Merge plate_df with controls table
+        plate_df = pd.merge(plate_df, controls, on='TubeCode', how='left')
+
+        # Assign sample_names ('Sample') to Katharoseq controls
+        plate_df['Sample'] = np.where(
+            (plate_df['Sample'].isna()) &
+            (plate_df == 'positive_control').any(axis=1),
+            "kathseq." + plate_df['number_of_cells'].astype(str) + "." +
+            plate_df['Row'] + plate_df['Col'].astype(str),
+            plate_df['Sample'])
+
+    else:
+
+        # Merge plate_df with controls table
+        plate_df = pd.merge(plate_df, blanks, on='TubeCode', how='left')
+
+    # Assign sample_names ('Sample') to BLANKS controls
+    plate_df['Sample'] = np.where(
+        (plate_df['Sample'].isna()) &
+        (plate_df == 'negative_control').any(axis=1),
+        "BLANK" + plate_df['Project Plate'].str.split('_').str.get(-1) +
+        "." + plate_df['Row'] + plate_df['Col'].astype(str),
+        plate_df['Sample'])
+
+    # Assign BLANK column
+    plate_df['Blank'] = np.where(plate_df['Sample'].str.contains('BLANK'),
+                                 True, False)
+
+    warnings.warn('Controls added')
+    n_blanks = plate_df['Blank'].sum()
+    if n_blanks < 6:
+        warnings.warn(f'There are only {n_blanks} in this prep. The'
+                      'recommended minimum number of blanks is 6')
+
+    return plate_df
+
+
+def validate_plate_df(plate_df, metadata, sample_accession_df, blanks_dir,
+                      katharoseq_dir=None):
+    """"Function checks that all the samples names recorded in the plate_df
+    have metadata associated with them. It also checks that all the matrix
+    tubes in the plate_df are indeed located in the sample accesion file or
+    controls lists.Checks for duplicate sample names and makes sure the
+    Katharoseq Dilution curve has the full dilution series (8-point serial
+    dilution)
+    ----------
+    Args:
+    plate_df:
+        pandas DataFrame object
+    metadata:
+        pandas DataFrame of qiita metadata associated with study
+    blanks dir: dir
+        "*.tsv" files of tube IDs assigned to blank tubes
+    katharoseq_dir: dir (default:None)
+        "*_tube_ids.tsv", contains tube ids of katharoseq samples
+        "*_cells_counts.tsv", contains cell counts of katharoseq samples
+
+    Returns
+    -------
+    Returns 0 (zero)
```

Comment:
it actually doesn't return anything, just raises errors or warnings as it goes through validation.